### ISSUE TITLE
With MSW example: fetch failed

### ISSUE THREAD

**Issue by thijssmudde** [vercel/next.js#44384]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

❯ npx --no-install next info

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64
    Binaries:
      Node: 19.3.0
      npm: 9.2.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.1.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

### Which example does this report relate to?

with-msw

### What browser are you using? (if relevant)

Chrome 108.0.5359.124 (Official Build) (x86_64)

### How are you deploying your application? (if relevant)

next dev

### Describe the Bug

TypeError: fetch failed

<img width="608" alt="image" src="https://user-images.githubusercontent.com/7371397/209718152-ea27c8a4-9b32-45fe-8991-1a451f71528e.png">

```
  cause: Error: getaddrinfo ENOTFOUND my.backend
      at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:107:26) {
    errno: -3008,
    code: 'ENOTFOUND',
    syscall: 'getaddrinfo',
    hostname: 'my.backend'
  },
  page: '/'
```

### Expected Behavior

The fetch should be handled by MSW. However, it seems the request is not resolved by MSW.

### To Reproduce

```
npx create-next-app --example with-msw with-msw-app
```

Then simply run yarn dev, and check the browser.

Or deploy to Vercel, it will show an error: https://with-msw-pearl.vercel.app/

**Comment by Razilator** [comment]
Hello, when building from a container, I get a similar error, I set it up for 3-4 hours, which I just didn’t change, the error remains ...
at the same time, the container with the backend exists and is available ...
```
GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:107:26) {
#0 23.34     errno: -3001,
#0 23.34     code: 'EAI_AGAIN',
#0 23.34     syscall: 'getaddrinfo',
#0 23.34     hostname: 'backend'
#0 23.34   }
#0 23.34 }
#0 23.35
#0 23.35 > Build error occurred
```

**Comment by Tarang74** [comment]
I'm receiving a similar error on my development server:
```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11118:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Page (webpack-internal:///(sc_server)/./app/review/page.tsx:13:22) {
  cause: Error: connect ECONNREFUSED ::1:8000
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1300:16)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -4078,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '::1',
    port: 8000
  }
}
```
The code referenced in `page.tsx` is the following:
```ts
export default async function Page() {
    const response = await fetch(/* ... */);
    ...
}
```

**Comment by farlandlee** [comment]
I am getting this issue as well.  Seems to be similar to something that @andykenward fixed previously?  I tried his change to targeting esNext and top level await, but to no avail.  I tried a few other demos online like this one, https://github.com/jmagrippis/eri, which is pinned to older versions of next, msw, etc and it fails as well.

Seems like it is related to the latest version of the macOS?  Are the rest of you also running the latest version?

> npx --no-install next info

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:03:51 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T6000
    Binaries:
      Node: 18.12.1
      npm: 8.19.2
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.1.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

**Comment by andykenward** [comment]
@fullhdpixel &  @farlandlee Have you tried a lower version of Node? I see you are using 18 or 19. Node 18 isn't supported by MSW yet, https://github.com/mswjs/msw/issues/1388 . So it's unlikely that Node 19 is either.

**Comment by farlandlee** [comment]
@andykenward you bet! Thanks for jumping in!

I just tried this configuration (node 16 and I updated next to canary)
```
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:03:51 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.18.1
      npm: 8.19.2
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.1.2-canary.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```
and I still get the same error.

In testing my operating system theory, I asked a co-worker who uses Windows to test it as well and his works, but only after he installed the canary version of Next.  Here is his setup:
```
Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 16.18.1
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.1.2-canary.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by farlandlee** [comment]
@fullhdpixel what happens if you run `npm install next@canary` does that solve the problem for you?  It looks like you are running on Intel and I have the M1 Pro chip, so maybe canary will solve it for you like it did my co-worker

**Comment by andykenward** [comment]
Using the [`with-msw`](https://github.com/vercel/next.js/tree/bd6a431b95966bce3cd505f8b360260bdffa8680/examples/with-msw) example 

`npx create-next-app@latest --example with-msw with-msw-app`

With
```
    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:03:51 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.18.1
      npm: 8.19.2
      Yarn: 1.22.19
      pnpm: 7.22.0
    Relevant packages:
      next: 13.1.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

The request works on the second client side request made in the browser. See the below screen recording.

https://user-images.githubusercontent.com/4893048/211056752-a5820a45-9f05-4787-abb4-b7d97d86da60.mp4

The issue of it not working on the first client side request is what I solved in this [comment](https://github.com/mswjs/msw/issues/1340#issuecomment-1324006370). There is a repo referenced there if you want to try it out.

I also tried `13.1.2-canary.1` version of next and it did the same as `13.1.1`. When using Node `16.18.1`.

When i switched to using Node `18.13.0`. I got the original issue filed by @fullhdpixel . See this great [comment](https://github.com/mswjs/msw/issues/1388#issuecomment-1241180921)  about MSW support for Node 18 native fetch support issue.

![CleanShot 2023-01-06 at 16 48 16](https://user-images.githubusercontent.com/4893048/211058655-d4092542-cbc0-4be7-88b1-3264af2e99e1.png)

@farlandlee I'm not sure it's an OS related issue. You might try checking that the MSW service worker is updated in the browser via the devtools. And also try deleting the `.next` & `node_modules` folder in your project if you've been changing node versions while trying to solve your issue.

**Comment by farlandlee** [comment]
@andykenward thanks again.  You are correct.  I missed that msw doesn't support node 18 and I had not changed the version.  But when I run the app on node 16, it does, indeed, load on a refresh of the page, just not on the originally load.  I had previously tried your solution for getting it to work on the first load, but since it did not work (due to my using node 18), I thought that was a dead end and rolled back.  I will make your recommended changes again and try running it on node 16 and see if the problem goes away.  I really appreciate your help!

**Comment by farlandlee** [comment]
@andykenward You were right...again :). I made your changes to be able to handle the first load and it is working for me on my work computer with this setup:
npx --no-install next info
```
    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:03:51 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.18.1
      npm: 8.19.2
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.1.2-canary.1
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```
Thanks again @andykenward! I owe you a pint.  🍺

**Comment by SylRob** [comment]
Hi,
the issue seem to have a link with the latest version of `node-fetch` and its use of ESM only [MSW Issue](https://github.com/mswjs/msw/issues/1388#issuecomment-1401706972).
I personally chose to ask Node to use its previous version of fetch by doing something like this :
```
"scripts": {
  "dev": "NODE_OPTIONS='--no-experimental-fetch' next dev",
  ...
}
```
(I'm on Node v18.13.0 BTW)

**Comment by pmuens** [comment]
I ran into this exact issue (using Node v19.8.1) and the [solution](https://github.com/vercel/next.js/issues/44384#issuecomment-1420476175) posted by @SylRob (using `NODE_OPTIONS`) fixed it for me.

> Hi, the issue seem to have a link with the latest version of `node-fetch` and its use of ESM only [MSW Issue](https://github.com/mswjs/msw/issues/1388#issuecomment-1401706972). I personally chose to ask Node to use its previous version of fetch (as a temporary fix) by doing something like this :
> 
> ```
> "scripts": {
>   "dev": "NODE_OPTIONS='--no-experimental-fetch' next dev",
>   ...
> }
> ```
> 
> (I'm on Node v18.13.0 BTW)

**Comment by Geminii** [comment]
Testing with the provided example : 

```
    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:21:34 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T8112
    Binaries:
      Node: 18.12.1
      npm: 8.19.2
      Yarn: N/A
      pnpm: 8.6.2
    Relevant Packages:
      next: 13.4.11
      eslint-config-next: 13.4.11
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

Tested with `"dev": "NODE_OPTIONS='--no-experimental-fetch' next dev"`  and not working 😭

**Comment by khoi-fish** [comment]
Hey guys. I've opened a PR for a fix: https://github.com/vercel/next.js/pull/54031

Note that while this will fix the issue described here, there is still another issue with `msw` not intercepting requests after a refresh. https://github.com/mswjs/msw/issues/1698

**Comment by kristojorg** [comment]
I was also only able to solve this by pinning Next `13.6.1` and using `"dev": "NODE_OPTIONS='--no-experimental-fetch' next dev"`

**Comment by balazsorban44** [comment]
Closing, as it's the same as https://github.com/vercel/next.js/issues/56608#issuecomment-1752923860
